includes/class-wc-payments-payment-request-button-handler.php

```diff
@@ -1700,4 +1700,44 @@ private function get_taxes( $product, $price ) {
 		// Normally there should be a single tax, but `calc_tax` returns an array, let's use it.
 		return WC_Tax::calc_tax( $price, $rates, false );
 	}
+
+	/**
+	 * Returns true if the given product is a subscription that has free trial and requires shipping.
+	 * This could be a subscription product with a trial period or a synchronised subscription with a delayed payment.
+	 *
+	 * If the product is a variable subscription, this function will return true if any of its variations have a trial and require shipping.
+	 *
+	 * @since 6.8.0
+	 *
+	 * @param WC_Product|null $product Product object.
+	 *
+	 * @return boolean
+	 */
+	public function product_has_trial_and_needs_shipping( $product ) {
+		if ( ! class_exists( 'WC_Subscriptions_Product' ) || ! class_exists( 'WC_Subscriptions_Synchroniser' ) || ! WC_Subscriptions_Product::is_subscription( $product ) ) {
+			return false;
+		}
+
+		if ( $product->get_type() === 'variable-subscription' ) {
+			$products = $product->get_available_variations( 'object' );
+		} else {
+			$products = [ $product ];
+		}
+
+		foreach ( $products as $product ) {
+			// Skip any products that are virtual as we only care about products that require shipping.
+			if ( ! $product->needs_shipping() ) {
+				continue;
+			}
+
+			// If the product has a trial period or is synchronised and the first payment is not today.
+			if ( WC_Subscriptions_Product::get_trial_length( $product ) > 0 ) {
+				return true;
+			} elseif ( WC_Subscriptions_Synchroniser::is_product_synced( $product ) && ! WC_Subscriptions_Synchroniser::is_payment_upfront( $product ) && ! WC_Subscriptions_Synchroniser::is_today( WC_Subscriptions_Synchroniser::calculate_first_payment_date( $product, 'timestamp' ) ) ) {
+				return true;
+			}
+		}
+
+		return false;
+	}
```

Comment:
@mattallan & I had a look at this and think there may be a less heavy handed approach to fix this issue. This function currently makes the whole variable product unpurchasable using express payment methods if 1 of the variations requires shipping. 

We agreed that the better way to go would be to use the `ajax_get_selected_product_data()` function or similar to limit the express payment method to eligible products. 

ie:

1. Change this `product_has_trial_and_needs_shipping()` function (and possibly its name) to check if **all** variations require shipping and have a free trial. 
2. Change `ajax_get_selected_product_data()` to throw an exception if the product is ineligible. 

That would have the effect of only loading the express payment method buttons if there is at least 1 eligible variation and hiding the express payment buttons if you choose a variation which is ineligible.

Reply:
Asking out of curiosity, do you have any type of product in mind that may have some variations that need shipping and some that do not need shipping? I did not think of this scenario initially :thinking:

Reply:
It could be like a variable product where a customer can sign up to the virtual copy or a physical copy of a magazine. But because this logic includes shipping and free trial this could also apply to stores who offers a free trial option on a variable product. eg subscribe now or sign up to our free trial.